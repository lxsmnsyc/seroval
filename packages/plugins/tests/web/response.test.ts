import {
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serializeAsync,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSONAsync,
} from 'seroval';
import { describe, expect, it } from 'vitest';
import ResponsePlugin from '../../web/response';

const EXAMPLE_BODY = 'Hello World!';

describe('Response', () => {
  describe('serializeAsync', () => {
    it('supports Response', async () => {
      const example = new Response(EXAMPLE_BODY);
      const result = await serializeAsync(example, {
        plugins: [ResponsePlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof example>(result);
      expect(back).toBeInstanceOf(Response);
      expect(await back.text()).toBe(await example.text());
    });
  });
  describe('toJSONAsync', () => {
    it('supports Response', async () => {
      const example = new Response(EXAMPLE_BODY);
      const result = await toJSONAsync(example, {
        plugins: [ResponsePlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof example>(result, {
        plugins: [ResponsePlugin],
      });
      expect(back).toBeInstanceOf(Response);
      expect(await back.text()).toBe(await example.text());
    });

    it('supports already read Response', async () => {
      const example = new Response(EXAMPLE_BODY);
      await example.text();
      const result = await toJSONAsync(example, {
        plugins: [ResponsePlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof example>(result, {
        plugins: [ResponsePlugin],
      });
      expect(back).toBeInstanceOf(Response);
      expect(back.body).toBe(null);
    });
  });

  describe('crossSerializeAsync', () => {
    it('supports Response', async () => {
      const example = new Response(EXAMPLE_BODY);
      const result = await crossSerializeAsync(example, {
        plugins: [ResponsePlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports Response', async () => {
        const example = new Response(EXAMPLE_BODY);
        const result = await crossSerializeAsync(example, {
          plugins: [ResponsePlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('crossSerializeStream', () => {
    it('supports Response', async () =>
      new Promise<void>((resolve, reject) => {
        const example = new Response(EXAMPLE_BODY);
        crossSerializeStream(example, {
          plugins: [ResponsePlugin],
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));

    it('supports already read Response', async () => {
      const example = new Response(EXAMPLE_BODY);
      await example.text();

      new Promise<void>((resolve, reject) => {
        crossSerializeStream(example, {
          plugins: [ResponsePlugin],
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      });
    });

    describe('scoped', () => {
      it('supports Response', async () =>
        new Promise<void>((resolve, reject) => {
          const example = new Response(EXAMPLE_BODY);
          crossSerializeStream(example, {
            plugins: [ResponsePlugin],
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toJSONAsync', () => {
    it('supports Response', async () => {
      const example = new Response(EXAMPLE_BODY);
      const result = await toCrossJSONAsync(example, {
        plugins: [ResponsePlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof example>(result, {
        plugins: [ResponsePlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(Response);
      expect(await back.text()).toBe(await example.text());
    });
  });

  it('supports already read Response', async () => {
    const example = new Response(EXAMPLE_BODY);
    await example.text();
    const result = await toCrossJSONAsync(example, {
      plugins: [ResponsePlugin],
    });
    expect(JSON.stringify(result)).toMatchSnapshot();
    const back = fromCrossJSON<typeof example>(result, {
      plugins: [ResponsePlugin],
      refs: new Map(),
    });
    expect(back).toBeInstanceOf(Response);
    expect(back.body).toBe(null);
  });

  describe('toCrossJSONStream', () => {
    it('supports Response', async () =>
      new Promise<void>((resolve, reject) => {
        const example = new Response(EXAMPLE_BODY);
        toCrossJSONStream(example, {
          plugins: [ResponsePlugin],
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
