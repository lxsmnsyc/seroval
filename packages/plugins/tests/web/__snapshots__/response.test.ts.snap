// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Response > crossSerializeAsync > scoped > supports Response 1`] = `"($R=>$R[0]=new Response($R[1]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,$R[2]={headers:$R[3]=new Headers($R[4]=[$R[5]=["content-type","text/plain;charset=UTF-8"]]),status:200,statusText:""}))($R["example"])"`;

exports[`Response > crossSerializeAsync > supports Response 1`] = `"$R[0]=new Response($R[1]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,$R[2]={headers:$R[3]=new Headers($R[4]=[$R[5]=["content-type","text/plain;charset=UTF-8"]]),status:200,statusText:""})"`;

exports[`Response > crossSerializeStream > scoped > supports Response 1`] = `"($R=>$R[0]=new Response($R[1]=($R[2]=d=>new ReadableStream({start:c=>{d.on({next:v=>{c.enqueue(v)},throw:v=>{c.error(v)},return:()=>{c.close()}})}}))($R[3]=($R[4]=(b,a,s,l,p,f,e,n)=>(b=[],a=!0,s=!1,l=[],p=0,f=(v,m,x)=>{for(x=0;x<p;x++)l[x]&&l[x][m](v)},n=(o,x,z,c)=>{for(x=0,z=b.length;x<z;x++)(c=b[x],(!a&&x===z-1)?o[s?"return":"throw"](c):o.next(c))},e=(o,t)=>(a&&(l[t=p++]=o),n(o),()=>{a&&(l[t]=void 0)}),{__SEROVAL_STREAM__:!0,on:o=>e(o),next:v=>{a&&(b.push(v),f(v,"next"))},throw:v=>{a&&(b.push(v),f(v,"throw"),a=s=!1,l.length=0)},return:v=>{a&&(b.push(v),f(v,"return"),a=!1,s=!0,l.length=0)}}))()),$R[5]={headers:$R[6]=new Headers($R[7]=[$R[8]=["content-type","text/plain;charset=UTF-8"]]),status:200,statusText:""}))($R["example"])"`;

exports[`Response > crossSerializeStream > scoped > supports Response 2`] = `"($R=>$R[3].next($R[9]=new Uint8Array($R[10]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,0,12)))($R["example"])"`;

exports[`Response > crossSerializeStream > scoped > supports Response 3`] = `"($R=>$R[3].return(void 0))($R["example"])"`;

exports[`Response > crossSerializeStream > supports Response 1`] = `"$R[0]=new Response($R[1]=($R[2]=d=>new ReadableStream({start:c=>{d.on({next:v=>{c.enqueue(v)},throw:v=>{c.error(v)},return:()=>{c.close()}})}}))($R[3]=($R[4]=(b,a,s,l,p,f,e,n)=>(b=[],a=!0,s=!1,l=[],p=0,f=(v,m,x)=>{for(x=0;x<p;x++)l[x]&&l[x][m](v)},n=(o,x,z,c)=>{for(x=0,z=b.length;x<z;x++)(c=b[x],(!a&&x===z-1)?o[s?"return":"throw"](c):o.next(c))},e=(o,t)=>(a&&(l[t=p++]=o),n(o),()=>{a&&(l[t]=void 0)}),{__SEROVAL_STREAM__:!0,on:o=>e(o),next:v=>{a&&(b.push(v),f(v,"next"))},throw:v=>{a&&(b.push(v),f(v,"throw"),a=s=!1,l.length=0)},return:v=>{a&&(b.push(v),f(v,"return"),a=!1,s=!0,l.length=0)}}))()),$R[5]={headers:$R[6]=new Headers($R[7]=[$R[8]=["content-type","text/plain;charset=UTF-8"]]),status:200,statusText:""})"`;

exports[`Response > crossSerializeStream > supports Response 2`] = `"$R[3].next($R[9]=new Uint8Array($R[10]=new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,0,12))"`;

exports[`Response > crossSerializeStream > supports Response 3`] = `"$R[3].return(void 0)"`;

exports[`Response > crossSerializeStream > supports already read Response 1`] = `"$R[0]=new Response(null,$R[1]={headers:$R[2]=new Headers($R[3]=[$R[4]=["content-type","text/plain;charset=UTF-8"]]),status:200,statusText:""})"`;

exports[`Response > serializeAsync > supports Response 1`] = `"new Response(new Uint8Array([72,101,108,108,111,32,87,111,114,108,100,33]).buffer,{headers:new Headers([["content-type","text/plain;charset=UTF-8"]]),status:200,statusText:""})"`;

exports[`Response > supports already read Response 1`] = `"{"t":25,"i":0,"s":{"body":{"t":2,"s":0},"options":{"t":10,"i":1,"p":{"k":["headers","status","statusText"],"v":[{"t":25,"i":2,"s":{"t":9,"i":3,"l":1,"a":[{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":0,"s":200},{"t":1,"s":""}],"s":3},"o":0}},"c":"seroval-plugins/web/Response"}"`;

exports[`Response > toCrossJSONStream > supports Response 1`] = `"{"t":25,"i":0,"s":{"body":{"t":25,"i":1,"s":{"factory":{"t":25,"i":2,"c":"seroval-plugins/web/ReadableStreamFactory"},"stream":{"t":31,"i":3,"a":[],"f":{"t":26,"i":4,"s":4}}},"c":"seroval/plugins/web/ReadableStream"},"options":{"t":10,"i":5,"p":{"k":["headers","status","statusText"],"v":[{"t":25,"i":6,"s":{"t":9,"i":7,"l":1,"a":[{"t":9,"i":8,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":0,"s":200},{"t":1,"s":""}],"s":3},"o":0}},"c":"seroval-plugins/web/Response"}"`;

exports[`Response > toCrossJSONStream > supports Response 2`] = `"{"t":32,"i":3,"f":{"t":15,"i":9,"l":12,"c":"Uint8Array","f":{"t":19,"i":10,"s":[72,101,108,108,111,32,87,111,114,108,100,33]},"b":0}}"`;

exports[`Response > toCrossJSONStream > supports Response 3`] = `"{"t":34,"i":3,"f":{"t":2,"s":1}}"`;

exports[`Response > toJSONAsync > supports Response 1`] = `"{"t":{"t":25,"i":0,"s":{"body":{"t":19,"i":1,"s":[72,101,108,108,111,32,87,111,114,108,100,33]},"options":{"t":10,"i":2,"p":{"k":["headers","status","statusText"],"v":[{"t":25,"i":3,"s":{"t":9,"i":4,"l":1,"a":[{"t":9,"i":5,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":0,"s":200},{"t":1,"s":""}],"s":3},"o":0}},"c":"seroval-plugins/web/Response"},"f":31,"m":[]}"`;

exports[`Response > toJSONAsync > supports Response 2`] = `"{"t":25,"i":0,"s":{"body":{"t":19,"i":1,"s":[72,101,108,108,111,32,87,111,114,108,100,33]},"options":{"t":10,"i":2,"p":{"k":["headers","status","statusText"],"v":[{"t":25,"i":3,"s":{"t":9,"i":4,"l":1,"a":[{"t":9,"i":5,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":0,"s":200},{"t":1,"s":""}],"s":3},"o":0}},"c":"seroval-plugins/web/Response"}"`;

exports[`Response > toJSONAsync > supports already read Response 1`] = `"{"t":{"t":25,"i":0,"s":{"body":{"t":2,"s":0},"options":{"t":10,"i":1,"p":{"k":["headers","status","statusText"],"v":[{"t":25,"i":2,"s":{"t":9,"i":3,"l":1,"a":[{"t":9,"i":4,"l":2,"a":[{"t":1,"s":"content-type"},{"t":1,"s":"text/plain;charset=UTF-8"}],"o":0}],"o":0},"c":"seroval-plugins/web/Headers"},{"t":0,"s":200},{"t":1,"s":""}],"s":3},"o":0}},"c":"seroval-plugins/web/Response"},"f":31,"m":[]}"`;
